src/Interpreters/MergeJoin.cpp

```diff
@@ -465,6 +466,7 @@ MergeJoin::MergeJoin(std::shared_ptr<TableJoin> table_join_, const Block & right
 
     table_join->splitAdditionalColumns(right_sample_block, right_table_keys, right_columns_to_add);
     JoinCommon::removeLowCardinalityInplace(right_table_keys);
```

Comment:
Ok but I don't understand why any of these needed.

Reply:
We are removing low cardinality for left block
https://github.com/ClickHouse/ClickHouse/blob/c3a9cbe01c1c07d21cd1f89b31291a194fc21b25/src/Interpreters/MergeJoin.cpp#L611

But left and right block expected to have same type because `insertFrom` and `compareAt` methods called on them.
https://github.com/ClickHouse/ClickHouse/blob/c3a9cbe01c1c07d21cd1f89b31291a194fc21b25/src/Interpreters/MergeJoin.cpp#L91
https://github.com/ClickHouse/ClickHouse/blob/c3a9cbe01c1c07d21cd1f89b31291a194fc21b25/src/Interpreters/MergeJoin.cpp#L76

(actually `removeLowCardinality` for right blocks is performed too 
https://github.com/ClickHouse/ClickHouse/blob/c3a9cbe01c1c07d21cd1f89b31291a194fc21b25/src/Interpreters/MergeJoin.cpp#L590-L595

It's unclear why it works, because `disk_writer` gets _header with `LowCardinallity`_ and _BlockStream without `LowCardinallity`_. But after read we get _block with `LowCardinallity`_ and it leads to errors in methods described above

https://github.com/ClickHouse/ClickHouse/blob/c3a9cbe01c1c07d21cd1f89b31291a194fc21b25/src/Interpreters/MergeJoin.cpp#L945-L948 )

Reply:
But why it would not work if we don't remove LowCardinality neither from left and right blocks?
I thought that writing LowCardinality columns to disk can be more efficient...